### PR TITLE
Move database to docker volume instead of in ~/.ddev, fixes #714

### LINKED ISSUE #714: Work around database storage problems on Docker for Windows/Mac/Docker Toolbox
## What happened (or feature request):

MariaDB in a Debian-based container on Docker for Windows with a host-mounted volume for /var/lib/mysql fails to start (has trouble writing the data files to disk on the network mount).

MariaDB is also failing in new and fun ways with Docker Toolbox, which is the path toward supporting Windows 10 Home + older Macs.

To get around this, I propose that we do the following:

* Stop using host mounts for /var/lib/mysql and use Docker volumes instead
* When a database container is stopped, dump the database to a host mount (`mysqldump` or similar)
* When a database container is started, if the dump file is present, import it and delete it

The last two points would be sort of optional, but very convenient. We could also not include them by default and document how to set up that behavior in the app's hooks.

This sidesteps the issue of keeping the mysql data directory on a network mount, and also allows for persisting the database contents on the host. It also allows us to not dive into mariadb's source and figure out complex problems that we don't have the time to fully understand.

### PR DESCRIPTION
**Pull request by rfay**
## The Problem/Issue/Bug:

OP #714 - We've had no end of trouble bind-mounting the database onto the host on various platforms, especially Windows. 

## How this PR Solves The Problem:

* Switch to using docker volume
* Provide database snapshot technique, including snapshots on ddev rm -R

TODO:
- [x] Command to create a snapshot
- [x] Command to restore a snapshot
- [x] Option on --remove-data to create snapshot (or omit it)
- [x] Figure out how to migrate existing databases into a volume.
- [x] Test scripts should delete all ddev volumes before starting.
- [x] Too much text output when running mariabackup (creating snapshot)
- [x] mariadb container tests need to use volume, not bind mount.
- [x] Update minimum docker version to 18.06
- [x] Docs required
     - Update explanation of where db is stored
     - Update remove docs

## Manual Testing Instructions:

* **Before testing, please `docker pull drud/ddev-dbserver:20180801_volume`**
* **Before testing: Any docker-compose.*.yaml files need to be changed to `version: '3.6'`, since we were forced to do a version bump here. This is an important thing to point out in release notes and announcements.
* Sample d7 databases are provided at [d7tester_test1](https://github.com/rfay/ddev/raw/4b2a65150c29a6194779ac002a8ca32565766bca/pkg/ddevapp/testdata/revert_snapshot/d7tester_test_1.sql.gz) (one node, site name "d7tester 1") and [d7tester_test2](https://github.com/rfay/ddev/raw/4b2a65150c29a6194779ac002a8ca32565766bca/pkg/ddevapp/testdata/revert_snapshot/d7tester_test_2.sql.gz) (two nodes, site name "d7tester 2")
* Create a project, load a db
* Verify that the docker volume has been created with `docker volume ls`
* Use ddev remove --remove-data to remove the project, you should see a snapshot created and the volume destroyed.
* Test `ddev snapshot`, verify that the snapshot gets created.
* Load a different site (like d7tester_test2)
* Test `ddev revert-to-snapshot` and you should lose the change you made.
* Run `ddev rm` (with or without --remove-data) and you should see a snapshot created.
* Test migration. Create a project with ddev v1.0.0 and then do a `ddev start` with it using this version. Verify that the migration happens successfully and that you have the project you expected.

This may be sensitive to different OSs; although we have tests running on all supported OSs, it will be good to get manual testing experience with this on Windows and Linux (and WIndows 10 Home).

## Automated Testing Overview:

* TestDbMigration
* TestDdevRestoreSnapshot

And of course every database-related action tests the behavior.

Required tests:
- [x] snapshot creation
- [x] snapshot loading/reverting
- [x] Full test suite needs to be run at least manually on macOS with docker edge.

## Related Issue Link(s):

OP #714 
[Mariadb upgrade to 10.2 #947](https://github.com/drud/ddev/issues/947)

## Questions for discussion

- [ ]  Currently, `ddev rm` always creates a snapshot unless you do `ddev remove --create-snapshot=false`. Think that's the right behavior?
- [ ] It may be worth figuring out how to automatically update people's docker-compose.*.yaml to `version: '3.6'`, but I'm afraid this would be a fair bit of dead code for a few people's short-term need.

### CODE REVIEW

**Comment by rfay** [comment]
*Something* has been regularly killing off docker on macOS, completely fatal. I'm wondering if it's this PR, and wondering if it's the volume usage (for the first time) that's doing it. There was also too much output from the snapshot save and load, but I think that's calmed down now.

**Comment by rickmanelius** [comment]
So another thought regarding snapshots. I really, really like this overall idea. I know it's somewhat pressured to come in as a result of docker volume usage, so I'm hoping my feedback doesn't add too much scope. However, in an "ideal" use case I could see something akin to...

1. Prompting to Save a Snapshot During Removal
When running `ddev rm --remove-data`, then prompt the user with a question of whether or not they want to save a snapshot. Not all cases (particularly with testing) warrants saving snapshots that are not really valuable/necessary/wanted. Adding a flag "-f" would force this and bypass this prompt. And the default to make a snapshot would be a "y" so a quick hit of enter would allow them to save the snapshot. That might be overloading things a bit, but the default of always a snapshot seems like it could become overrun with backups that don't mean much.

2. Snapshot Generation & Restore
I can see this being something of great value, particularly when stepping through database changes as a site builder. I would see this more as 3 components:

A. Generation of backups
B. Query/list of backups.
C. Restore backups (default to last backup and, if running a query, pick one).

This could look something like...

* `ddev snapshot` => backups_dir/projectname_timestamp1
* `ddev snapshot` => backups_dir/projectname_timestamp2
* `ddev snapshot --list` => shows list.
* `ddev restore-snapshot` => restores last snapshot
* `ddev restore-snapshot timestamp` => restores specific snapshot

It's worth noting that if someone ran the list command, they could be prompted to enter a number to restore a particular backup or 0/n to not restore.

A complication is the fact that we also have snapshots from provider plugins (currently, only Pantheon), which stores in yet another archive format of an archive per type (db and files). This is where I was pushing to have a consistent file format that we saved archives in to somewhat marry  the two versions. Alas, we can still have them work as independent items for now. And the ability to query and revert to another backup is not needed out of the gates. But if we are storing multiple backups per project, we need a way to actually use that...

**Comment by rickmanelius** [comment]
Hi @rfay. I'm very excited about this feature, so of course I'll give it another test :)

Agreed that we can improve long term. My comments were more what/how I see this feature growing into and not something that had to be done right now. So yes, once we can validate a working snapshot revert (which I wasn't able to do my last go around), I vote to approve and we can file a subsequent wishlist item.

FWIW, here was one area where snapshots (particularly in the UI) was a conceived feature https://github.com/drud/ddev-ui/wiki/Roadmap#v050-workflow-improvements. This particularly true for the WP community, which tends to have more site building and the ability to "save state" and test some new changes with the ability to revert would be super valuable.

**Comment by rickmanelius** [comment]
Thanks for the improved testing instructions in the issue summary!

**Comment by rickmanelius** [comment]
ion

### DDEV Binary

* https://github.com/drud/ddev/pull/987
* `git checkout 08296b5f`
* `make darwin`
* `ddev version` => v0.20.0-38-g08296b5f (matches)
* docker pull drud/mariadb-local:20180711_volume


### Site

* Clean Up
  * Remove previous Drupal 7.59 site (ddev rm --remove-data && rm -rf)
* Download Drupal 7.59
* Unzip
* `ddev config --projectname drupal-7.59 --projecttype drupal7 --docroot .`
* `ddev start`

## Testing

* Verify Docker Volume Exists
  * docker volume ls shows a volume name "drupal-7.59-mariadb"
* Verify Snapshot Generation
  * Import db1 with `ddev import-db --src ~/Downloads/d7tester_test_1.sql.gz`
  * Verified 1 node with sitename "d7 tester test 1"
  * ddev remove --remove-data creates a snapshot in .ddev/db_snapshots/
  * Attempt to run `ddev snapshot` while in a stopped state.
    * I see lots of warnings and errors here. A simple "containers need to be operational or healthy" with a prompt to run `ddev start` would probably be better.
    * Despite the error, I did see another snapshot attempt based on the folder. The folder is in fact empty, but it's there.
* Verify a revert
  * From nothing, re-run ddev start to bring up the containers.
  * Import db 2 from `ddev import-db --src ~/Downloads/d7tester_test_2.sql.gz`
  * Verified two nodes and site name "d7 tester test 2"
  * Ran `ddev revert-to-snapshot`
  * I still see the 2nd database...
  * Running another `ddev rm --remove-data` and verified another snapshot created
  * Run `ddev start` and verified no db
  * Re-run `ddev revert-to-snapshot`
  * Still no database...

Looks like I'm not getting the revert working...

**Comment by rfay** [comment]
Well, I never would have thought of anybody trying to snapshot when it's not running, but that's why it's nice to have multiple eyes on these things. I

* Made it work if stopped or rm'd
* Made import-db work when stopped or rm'd, since that's always been an annoyance ("Why doesn't it just start itself?").

You should know that revert is reverting to the entire mariadb environment (mysql database, empty db database). So if you snapshot a beginner db, and then revert to it... you'll have a db ready to be installed. It's *not* just snapshotting the 'db' database, the the entire system. It wasn't clear from your writeup whether maybe that happened to you (snapshot and restore an empty database, with just the mysql db installed).

I'd love to walk through this with you, to try to completely understand the issue that's happening. Great to have multiple eyes on something like this!

**Comment by rfay** [comment]
I have good news and bad news.
* Good news: upstream/master does *not* seem to destroy docker on macOS. I've run it manually 4 times and am starting a 20-cycle run of it. 
* Bad news: Testing this PR (just the buildkite test.sh) does destroy docker on macOS. I'm able to recreate complete docker failure after one or two runs on macOS. You get a dialog box that says "Fatal error: com.docker.supervisor failed to start. Exit code 1 "

So the bad news is 
* I don't think this can be a candidate for v1.0.0, because we have to have confidence that this behavior is stopped.
* Unfortunately there are a number of fixes and features in here that will be sad to leave out (or move in another way). Not just snapshotting (which should work without docker volumes, assuming docker volumes are the fragility here). But significant improvements to tests (including the ability to test content on a running ddev project) and test script improvements. Sad.

Changing the tag to v1.1.0, we'll see what happens..

**Comment by rfay** [comment]
The problem with this PR is that destroys docker, whether in general use or running tests. I've removed it from the upcoming release, and it will be a significant effort to chase what's happening. But what it does, if you test it (on macOS at least) is it brings down docker on the testbots, meaning that all our testing is completely destabilized and confusing.

To continue chasing I'll have to manually run tests against an early, early commit that has only the docker volume stuff; I'm pretty sure there was a clean one running using the volume technique. I can put the test script in a loop and run that hard on a disabled macOS testbot.  It's certainly possible that the volume approach is OK but something else here stresses Docker.

I'm fine with meeting and testing it tomorrow, but not sure it's worth your time, as this is not ready for release, and may not ever make it as is. It's passing automated tests that do everything I specified to do manually, so I think we'll be successful walking through those.  So it's up to you.

I just can't imagine the pain if we break every mac user's docker :(

**Comment by rfay** [comment]
Rick and I walked through this today and figured out what was happening! He was using `ddev revert-to-snapshot` and it accepted that! And did nothing... We decided on these outcomes:

1. ddev revert-to-snapshot needs to require the arg of which snapshot!
2. revert-to-snapshot might also (if no arg) display some snapshots for attention and say how to use it.
3. The args to ddev rm should be improved. Probably --create-snapshot and --omit-snapshot, with no binary arg like currently done. 
4. ddev rm --remove-data should do a snapshot by default.
5. Regular ddev rm show not do the snapshot by default.

Rick, I'll edit if I missed something there.

**Comment by rickmanelius** [comment]
Hi @rfay. Looks good! Thanks for summarizing. I think there was one more item, and I had an additional idea after our meeting.

* --remove-data => --remove-db to indicate that this isn't about "data" but rather specifically the database.
* idea: revert-to-snapshot is a bit long. What about "restore"? I agree "revert-to-snapshot" is more precise and we can always alias laster.
* I believe this summarizes the syntax if these suggestions work:
  * ddev rm --remove-db -O (ommits snapshot)
  * ddev rm --remove-db -S (creates snapshot)
  * ddev restore --snapshot-name=snapshot

**Comment by rfay** [comment]
@rickmanelius 
* on --remove-data => --remove-db: We've had --remove-data for a *really* long time; it's easy enough to add --remove-db as an alias, even make it primary, but we can't just change it. And remember we have a specific and valid request to actually do all the removal: https://github.com/drud/ddev/issues/989 - So I'm uncomfortable with this change and don't think it takes us where we want to go.

**Comment by rickmanelius** [comment]
@rfay 

1. I'm fine with an alias or simply keep `--remote-data`. We only suggested the addition of "database' or "db" to highlight that this isn't all data (i.e. code + files + db), but specifically the database. That said, nobody has complained and the alias is fine.
2. I have a question about the suggestion of `--restore-snapshot` as a flag. Do you mean this as a flag to `ddev snapshot` instead of creating a separate command? That is `ddev snapshot` to create and `ddev snapshot --restore --name=foldername`?

**Comment by rfay** [comment]
1. We now have `ddev snapshot`, `ddev rm --create-snapshot`, `ddev rm --remove-data --omit-snapshot`
2. I changed restore to `ddev restore-snapshot <snapshot_name>`; if you don't give the name, it tells you where to look.

**Comment by rfay** [comment]
Apparently the macOS docker-for-mac crashing problem is in https://github.com/docker/for-mac/issues/2765; if I use the experimental version there, everything is completely solid.

**Comment by rickmanelius** [comment]
Oh that sucks... it looks like we may have to wait until 18.09 :(

**Comment by rfay** [comment]
The harsh test against this PR seems to run much, much longer (and fail eventually differently) with 18.06.0 rc3 (edge), so I'm hopeful we may be able to pursue this when 18.06 goes stable.  But it's on the shelf for now.

**Comment by rfay** [comment]
Added `ddev snapshot --name="some_meaningful_name"` so you can name snapshots for easier restore. 

I think this is ready for review, would love to see about getting it in.

Since this is a major change in our db storage, we need a bit of a performance review, at least so we know a little bit about what the impact might be.

In addition, it would be wonderful to see if mariadb 10.2 will run in this now (on Windows).

**Comment by rfay** [comment]
Performance (macOS High Sierra, 2017 MBP 2.3GHz Core I5, 16GB)
* D8 drush si -y inside container with this PR takes 43s or 40s  (second try)  as compared to 46s or 49s with v1.0.0. I'm not sure that's a relevant difference, but that's at least in the right direction.
* D8 Web install with this PR takes about 1:45, with v1.0.0 about 1:36, there may be measurement differences.

So basically it's in the same range, we have no real performance issues to worry about at least given this test. 

I should note that we have a mariadb caching possibility turned off for Windows 10 Home; we may be able to undo that and gain a bit.

File: cmd/ddev/cmd/restore_snapshot.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import (
+	"github.com/drud/ddev/pkg/ddevapp"
+	"github.com/drud/ddev/pkg/util"
+	"github.com/spf13/cobra"
+	"os"
+)
+
+// DdevRestoreSnapshotCommand provides the ability to revert to a database snapshot
+var DdevRestoreSnapshotCommand = &cobra.Command{
+	Use:   "restore-snapshot projectname_HHHHMMDDHHMMSS",
```
[the remark this answers is not in the document]
Comment (rfay) [reply cmd/ddev/cmd/restore_snapshot.go]: I changed it to work that way.

**Comment by rickmanelius** [comment]
I'm starting to test this now. Right off the bat I had an older config.yml (probably just v1.0.0) where re-running ddev config and then a `ddev start` resulted in:

```
Starting drupal-7.59...
Migrating bind-mounted database in ~/.ddev to docker-volume mounted database
Failed to start drupal-7.59: Failed to migrate db from bind-mounted db: failed to run migrate_file_to_volume.sh, err=failed to create/start docker container ({drupal-7.59_migrate_volume 0xc4202005a0 0xc42023a400 <nil> <nil>}):no such image output=
```

I'm going to blow everything out, but toggling between an established working v1.0.0 project and back will be useful to test the upgrade path. This one already feels like it's going to be a doozy... notably on existing projects.

**Comment by rickmanelius** [comment]
Also, looks like (based on the config.yml generated from a `ddev config`) that the manual testing instructions should be `docker pull drud/ddev-dbserver:20180801_volume`. Note the change from "mariadb" to "dbserver" in the string.

**Comment by rickmanelius** [comment]
Sorry for 3 quick comments in a row. There were two other factors that existed in this repo that might have caused the previous issue. I had a docker-compose.memcached.yaml file as well as pre-existing .ddev/db_snapshots. Either of those two (from previous tests) could be problematic. I'm going to tear down to a clean D7 repo and start over.

**Comment by rfay** [comment]
@rickmanelius it doesn't run rm/start when creating a snapshot. 

It does run rm/start when *restoring* a snapshot, and the reason is the mechanism: At startup time we either use the existing db or create one from scratch. In the past, we always created from scratch using the db baked into the image (which sped up startup times quite a lot). Now, with this PR, we use the exact same code, but if we have a snapshot specified it creates from the snapshot, so the basic code remains the same. I'm sure there's more than one way to do this.

**Pull request merged**